src/cmd/build.js

```diff
@@ -30,31 +33,48 @@ export default function build(
     fileFilter = new FileFilter();
   }
 
-  return resolveManifest
-    .then((manifestData) =>
-      Promise.all([
-        prepareArtifactsDir(artifactsDir),
-        zipDir(sourceDir, {
-          filter: (...args) => fileFilter.wantFile(...args),
-        }),
-      ])
-      .then((results) => {
-        let [artifactsDir, buffer] = results;
-        let packageName = safeFileName(
-          `${manifestData.name}-${manifestData.version}.xpi`);
-        let extensionPath = path.join(artifactsDir, packageName);
-        let stream = createWriteStream(extensionPath);
-        let promisedStream = streamToPromise(stream);
-
-        stream.write(buffer, () => stream.end());
-
-        return promisedStream
-          .then(() => {
-            log.info(`Your web extension is ready: ${extensionPath}`);
-            return {extensionPath};
-          });
-      })
-    );
+  if (!createPackage) {
+    createPackage = () => {
+      return resolveManifest
+        .then((manifestData) => {
+          return zipDir(
+            sourceDir, {
+              filter: (...args) => fileFilter.wantFile(...args),
+            })
+            .then((buffer) => {
+              let packageName = safeFileName(
+                `${manifestData.name}-${manifestData.version}.xpi`);
+              let extensionPath = path.join(artifactsDir, packageName);
+              let stream = createWriteStream(extensionPath);
+              let promisedStream = streamToPromise(stream);
+
+              stream.write(buffer, () => stream.end());
+
+              return promisedStream
+                .then(() => {
+                  log.info(`Your web extension is ready: ${extensionPath}`);
+                  return {extensionPath};
+                });
+            });
+        });
+    };
+  }
+
+  return prepareArtifactsDir(artifactsDir)
+    .then(() => createPackage())
+    .then((result) => {
+      if (rebuildWhenSourceChanges) {
+        log.info('Rebuilding when files change...');
+        onSourceChange({
+          sourceDir, artifactsDir, onChange: () => {
+            return createPackage().catch((error) => {
+              throw error;
```

Comment:
Good idea. I originally thought the thrown error would get logged (like default Node behavior) but for some reason it does not! I will add logging.